ambari-metrics/ambari-metrics-kafka-sink/src/main/java/org/apache/hadoop/metrics2/sink/kafka/JvmMetricSet.java

```diff
@@ -0,0 +1,200 @@
+    this.threadMXBean = threadMXBean;
+    this.runtimeMXBean = runtimeMXBean;
+  }
+
+  public Map<MetricName, Gauge<?>> getJvmMetrics() {
+    return Stream.concat(
+      getMemoryUsageMetrics().stream(),
+      Stream.concat(
+        getThreadMetrics().stream(),
+        Stream.of(getRuntimeMetrics())
+      ))
+      .collect(Collectors.toMap(JvmMetric::getMetricName, JvmMetric::getMetric));
+  }
+
+  private List<JvmMetric> getMemoryUsageMetrics() {
+
+    return Stream.of(
+      new AbstractMap.SimpleEntry<>("heap_usage", memoryMXBean.getHeapMemoryUsage()),
+      new AbstractMap.SimpleEntry<>("non_heap_usage", memoryMXBean.getNonHeapMemoryUsage()))
+      .map(entry ->
+        new JvmMetric(
+          MetricNameBuilder.builder().type(MEMORY).name(entry.getKey()).build(),
+          new RatioGauge() {
+
+            @Override
+            protected double getNumerator() {
+              return entry.getValue().getUsed();
+            }
+
+            @Override
+            protected double getDenominator() {
+              return entry.getValue().getMax();
+            }
+          }
+        ))
+      .collect(Collectors.toList());
+
+  }
+
+  private List<JvmMetric> getThreadMetrics() {
+
+    return
+      Stream.concat(
+        Stream.of(
+          new JvmMetric(
+            MetricNameBuilder.builder().type(THREADS).name("thread_count").build(),
+            new Gauge<Integer>() {
+              @Override
+              public Integer value() {
+                return threadMXBean.getThreadCount();
+              }
+            }
+          ),
+          new JvmMetric(
+            MetricNameBuilder.builder().type(THREADS).name("daemon_thread_count").build(),
+            new Gauge<Integer>() {
+              @Override
+              public Integer value() {
+                return threadMXBean.getDaemonThreadCount();
+              }
+            }
+          )),
+        Stream
+          .of(Thread.State.RUNNABLE, Thread.State.BLOCKED, Thread.State.TIMED_WAITING, Thread.State.TERMINATED)
+          .map(state -> new JvmMetric(
+            MetricNameBuilder.builder().type(THREADS).name(getThreadMetricNameByState(state)).build(),
+            new Gauge<Long>() {
+              @Override
+              public Long value() {
+                return getThreadCountByState(state);
+              }
+            }
+          )))
+        .collect(Collectors.toList());
+  }
+
+  private String getThreadMetricNameByState(Thread.State state) {
+    String name = "thread-states.";
+    switch (state) {
+      case BLOCKED:
+        name += "blocked";
+        break;
+      case RUNNABLE:
+        name += "runnable";
+        break;
+      case TIMED_WAITING:
+        name += "timed_waiting";
+        break;
+      case TERMINATED:
+        name += "terminated";
+        break;
+      case NEW:
+        name += "new";
+        break;
+      case WAITING:
+        name += "waiting";
+        break;
+    }
+    return name;
```

Comment:
Why not replace this huge  case with hard-code to `state.name().toLowerCase()` ?

i.e. `return String.format("thread-states.%s", state.name().toLowerCase());`  with adding `@Nonnull` for state argument

Reply:
make sense